src/nestegg.c

```diff
@@ -1553,6 +1561,36 @@ ne_read_block(nestegg * ctx, uint64_t block_id, uint64_t block_size, nestegg_pac
         }
         f->frame_encryption->length = IV_SIZE;
         encryption_size = SIGNAL_BYTE_SIZE + IV_SIZE;
+
+        if ((signal_byte & PARTITIONED_BIT_MASK) == PACKET_PARTITIONED) {
+          r = ne_io_read(ctx->io, &f->frame_encryption->num_partitions, NUM_PACKETS_SIZE);
+          if (r != 1) {
+            free(f->frame_encryption->iv);
+            free(f->frame_encryption);
+            free(f);
+            nestegg_free_packet(pkt);
+            return r;
+          }
+          
+          encryption_size += NUM_PACKETS_SIZE + f->frame_encryption->num_partitions * PACKET_OFFSET_SIZE;
+          f->frame_encryption->partition_offsets = ne_alloc(f->frame_encryption->num_partitions * PACKET_OFFSET_SIZE);
+          r = 0;
+          for (i = 0; i < f->frame_encryption->num_partitions; i++) {
+            uint64_t value = 0;
+            r += ne_read_uint(ctx->io, &value, PACKET_OFFSET_SIZE);
+            f->frame_encryption->partition_offsets[i] = (uint32_t)value;
+          }
+
+          /* If any of the partition offsets did not return 1, then fail. */
+          if (r != f->frame_encryption->num_partitions) {
+            free(f->frame_encryption->iv);
+            free(f->frame_encryption);
+            free(f->frame_encryption->partition_offsets);
+            free(f);
+            nestegg_free_packet(pkt);
+            return r;
```

Comment:
The difference between `ne_read_uint` returning -1 and 0 as an error is important for the final result of `nestegg_read_packet` because it's used to differentiate between a short read (e.g. not enough data available to complete the parse) and a hard parse error, so you need to preserve that.

I think the simplest change would be to drop the add-to-`r` trick and instead break if `ne_read_uint` returns != 1.  You can use the value of `i` to determine if you've read all of the partition offsets.